// react-fix-image-orientation/lib/ExifOrientationImg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _exifJs = require("exif-js");

var _exifJs2 = _interopRequireDefault(_exifJs);

var _exif2css = require("exif2css");

var _exif2css2 = _interopRequireDefault(_exif2css);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function isMobile() {
  var check = false;
  var userAgent = window.navigator.userAgent || window.navigator.vendor || window.opera;
  if (/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(userAgent) || /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(userAgent.substr(0, 4))) check = true;
  return check;
}

var snakeToCamelCaseKeys = function snakeToCamelCaseKeys(obj) {
  return Object.keys(obj).map(function (k) {
    return _defineProperty({}, k.replace(/(-\w)/g, function (m) {
      return m[1].toUpperCase();
    }), obj[k]);
  }).reduce(function (a, b) {
    return _extends({}, a, b);
  }, {});
};

var ExifOrientationImg = function (_Component) {
  _inherits(ExifOrientationImg, _Component);

  function ExifOrientationImg() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, ExifOrientationImg);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = ExifOrientationImg.__proto__ || Object.getPrototypeOf(ExifOrientationImg)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      orientation: null
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ExifOrientationImg, [{
    key: "_onImageLoaded",
    value: function _onImageLoaded() {
      var _this2 = this;

      for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      var event = args[0],
          otherArgs = args.slice(1);

      var imageElement = event.target;
      var onLoad = this.props.onLoad;

      // Fix for an issue affecting exif-js: see https://github.com/exif-js/exif-js/issues/95

      var windowImage = window.Image;
      window.Image = null;

      // Do the actual EXIF operations[
      if (!_exifJs2.default.getData(imageElement, function () {
        !isMobile() && _this2.setState({
          orientation: _exifJs2.default.getTag(imageElement, "Orientation")
        });
        onLoad && onLoad.apply(undefined, [event].concat(_toConsumableArray(otherArgs)));
      })) {
        onLoad && onLoad.apply(undefined, [event].concat(_toConsumableArray(otherArgs)));
      }

      // Re-establish the reference
      window.Image = windowImage;
    }
  }, {
    key: "render",
    value: function render() {
      var _props = this.props,
          src = _props.src,
          alt = _props.alt,
          _props$style = _props.style,
          style = _props$style === undefined ? {} : _props$style,
          onLoad = _props.onLoad,
          innerRef = _props.innerRef,
          ref = _props.ref,
          imgProps = _objectWithoutProperties(_props, ["src", "alt", "style", "onLoad", "innerRef", "ref"]);

      var orientation = this.state.orientation;


      return _react2.default.createElement("img", _extends({
        onLoad: this._onImageLoaded.bind(this),
        src: src,
        alt: alt,
        ref: innerRef || ref,
        style: _extends({}, orientation ? snakeToCamelCaseKeys((0, _exif2css2.default)(orientation)) : {}, style)
      }, imgProps));
    }
  }]);

  return ExifOrientationImg;
}(_react.Component);

exports.default = ExifOrientationImg;
